### PR TITLE
Allow tagging of array axes in `ArrayContext.einsum`

### PR DESCRIPTION
**Pull request by thomasgibson**
An initial quick attempt to allow array axes tagging in `ArrayContext.einsum`.

### CODE REVIEW

File: meshmode/array_context.py [hunk 1]
```diff
@@ -424,13 +425,34 @@ def einsum(self, spec, *args, arg_names=None, tagged=()):
             generated.
         :arg tagged: an optional sequence of :class:`pytools.tag.Tag`
             objects specifying the tags to be applied to the operation.
+        :arg tagged_array_axes: an optional :class:`Dict` whose keys
+            correspond to specific *arg_names* with values consisting
+            of strings that parse to a
+            :class:`~loopy.kernel.array.ArrayDimImplementationTag`.
+            See :func:`~loopy.kernel.array.parse_array_dim_tags` for a
+            description of the allowed string format.
```
Comment (inducer) [inline meshmode/array_context.py:433]: Anything captured at this level should be
* descriptive ("this axis describes the element number"), not prescriptive ("use this memory layout for this axis"), i.e. a subclass of `pytools.tag.Tag`
* as a result, definitely not a string.

**Review by inducer — COMMENTED** [review]
Thanks for thinking about this. A comment about the approach below.

File: meshmode/array_context.py (hunk 1 above)
Comment (inducer) [reply meshmode/array_context.py:433]: (And, to be clear, the translation from "descriptive" to "prescriptive" should be done centrally in program transformation.

As a matter of justification: If the application layer (the bit that describes the math) gets involved in prescribing memory layouts, it takes freedom *away* from the transform layer to do what it wants. In that sense, prescriptive tags here would be a layering violation.

**Comment by thomasgibson** [comment]
Closing this since loopy kernels are already taggable

**Pull request closed without merging**